File: math/leastInterval.java
import java.util.Arrays;
import java.util.Scanner;

/*
	problem: Task Scheduler
	给定一个用字符数组表示的 CPU 需要执行的任务列表。其中包含使用大写的 A - Z 字母表示的26 种不同种类的任务。任务可以以任意顺序执行，并且每个任务都可以在 1 个单位时间内执行完。CPU 在任何一个单位时间内都可以执行一个任务，或者在待命状态。

	然而，两个相同种类的任务之间必须有长度为 n 的冷却时间，因此至少有连续 n 个单位时间内 CPU 在执行不同的任务，或者在待命状态。

	你需要计算完成所有任务所需要的最短时间。

	示例 1：

	输入: tasks = ["A","A","A","B","B","B"], n = 2
	输出: 8
	执行顺序: A -> B -> (待命) -> A -> B -> (待命) -> A -> B.
	注：

	任务的总个数为 [1, 10000]。
	n 的取值范围为 [0, 100]。

	idea:
	（1）因为每个任务的执行都需要一个单位时间，所以执行完全部任务，用时不会少于任务个数
	（2）其次，相同种类的任务之间有长度为n的冷却时间，所以要尽可能利用起来这些冷却时间，才能使总时间少
	（3）给定一个任务列表，规划任务执行顺序：
		a. 对每个任务的实例进行计数，同时初始化一个任务计时器time（初值为0）
		b. 按实例数排序（从小到大或者从大到小），首先要执行实例数最多的任务，计时器time递增
		c. 其次执行实例数次多的任务，计时器time递增，直到执行任务所用的时间达到冷却时长
		d. 重复b、c，直到实例数最多的任务个数为0，返回time


*/

public class Solution {
    public int leastInterval(char[] tasks, int n) {
        int[] map = new int[26];

        for (char c: tasks)
            map[c - 'A']++;
        Arrays.sort(map);

        int time = 0;
        while (map[25] > 0) {
            int i = 0;
            while (i <= n) {
                if (map[25] == 0)
                    break;
                if (i < 26 && map[25 - i] > 0)
                    map[25 - i]--;
                time++;
                i++;
            }
            Arrays.sort(map);
        }

        return time;
    }
}

Class LeastInterval{
	public static void main(String[] args) {
		char [] tasks = {'B', 'D', 'A', 'A', 'E', 'A', 'A', 'C', 'A'};

		Scanner scan = new Scanner(System.in);
		int n = scan.nextInt();

		Solution solu = new Solution();
		solu.leastInterval(tasks, n);
	}
}

File: math/leastInterval.cpp
/*
problem:
Given a char array representing tasks CPU need to do. It contains capital letters A to Z where different letters represent different tasks. Tasks could be done without original order. Each task could be done in one interval. For each interval, CPU could finish one task or just be idle.

However, there is a non-negative cooling interval n that means between two same tasks, there must be at least n intervals that CPU are doing different tasks or just be idle.

You need to return the least number of intervals the CPU will take to finish all the given tasks.

 

Example:

Input: tasks = ["A","A","A","B","B","B"], n = 2
Output: 8
Explanation: A -> B -> idle -> A -> B -> idle -> A -> B.
 

Note:

The number of tasks is in the range [1, 10000].
The integer n is in the range [0, 100].

idea:
reference: https://leetcode.com/articles/task-scheduler/
    solution one:   用排序的方法做，
                    1. 在小于冷却时间的cycle内，每次选次数最大task执行，计时
                    2. 超过冷却时间cycle，根据每个task剩余的次数，对task排序
                    3. 重复1，2，直到所有task剩余数目为0，其实相当于次数最多的task为0
Before we start off with the solution, we can note that the names of the tasks are irrelevant for obtaining the solution of the given problem. The time taken for the tasks to be finished is only dependent on the number of instances of each task and not on the names of tasks.

The first solution that comes to the mind is to consider the tasks to be executed in the descending order of their number of instances. For every task executed, we can keep a track of the time at which this task was executed in order to consider the impact of cooling time in the future. We can execute all the tasks in the descending order of their number of instances and can keep on updating the number of instances pending for each task as well. After one cycle of the task list is executed, we can again start with the first task(largest count of instances) and keep on continuing the process by inserting idle cycles wherever appropriate by considering the last execution time of the task and the cooling time as well.

But, there is a flaw in the above idea. Consider the case, where say the number of instances of tasks A, B, C, D, E are 6, 1, 1, 1, 1 respectively with n=2(cooling time). If we go by the above method, firstly we give 1 round to each A, B, C, D and E. Now, only 5 instances of A are pending, but each instance will take 3 time units to complete because of cooling time. But a better way to schedule the tasks will be this: A, B, C, A, D, E, ... . In this way, by giving turn to the task A as soon as its cooling time is over, we can save a good number of clock cycles.

From the above example, we are clear with one idea. It is that, the tasks with the currently maximum number of outstanding (pending)instances will contribute to a large number of idle cycles in the future, if not executed with appropriate interleavings with the other tasks. Thus, we need to re-execute such a task as soon as its cooling time is finished.

Thus, based on the above ideas, firstly, we obtain a count of the number of instances of each task in mapmap array. Then, we start executing the tasks in the order of descending number of their initial instances. As soon as we execute the first task, we start its cooling timer as well(ii). For every task executed, we update the pending number of instances of the current task. We update the current time, timetime, at every instant as well. Now, as soon as the timer, ii's value exceeds the cooling time, as discussed above, we again need to consider the task with the largest number of pending instances. Thus, we again sort the taskstasks array with updated counts of instances and again pick up the tasks in the descending order of their number of instances.

Now, the task picked up first after the sorting, will either be the first task picked up in the last iteration(which will now be picked after its cooling time has been finished) or the task picked will be the one which lies at (n+1)^{th}(n+1) 
th
  position in the previous descending taskstasks array. In either of the cases, the cooling time won't cause any conflicts(it has been considered implicitly). Further, the task most critical currently will always be picked up which was the main requirement.

We stop this process, when the pending instances of all the tasks have been reduced to 0. At this moment, timetime gives the required result.
*/
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        vector<int> task2num(26, 0);
        for (auto task: tasks)
            task2num[task - 'A']++;

        sort(task2num.begin(), task2num.end()); // 默认从小到大排，所以排序之后，task2num[25] 的值最大

        int time = 0;
        while (task2num[25] > 0) {  // 表明 task 还没执行完
            int i = 0;
            while (i <= n) {
                if (task2num[25] == 0)
                    break;
                if (i < 26 && task2num[25 - i] > 0)
                    task2num[25 - i]--;
                time++;
                i++;
            }
            sort(task2num.begin(), task2num.end());
        }
        vector<int> ().swap(task2num);  // 释放向量空间
        return time;
    }
};

/*
涉及到没次取最大值，就应该想到 “优先队列”
这个版本是 Java 的，已经 Accepted 了；C++ 出队，个数减1，不知道还是优先队列吗？总之C++这个版本没实现出来
public class Solution {
    public int leastInterval(char[] tasks, int n) {
        int[] map = new int[26];
        for (char c: tasks)
            map[c - 'A']++;
        PriorityQueue < Integer > queue = new PriorityQueue < > (26, Collections.reverseOrder());
        for (int f: map) {
            if (f > 0)
                queue.add(f);
        }
        int time = 0;
        while (!queue.isEmpty()) {
            int i = 0;
            List < Integer > temp = new ArrayList < > ();
            while (i <= n) {
                if (!queue.isEmpty()) {
                    if (queue.peek() > 1)
                        temp.add(queue.poll() - 1);
                    else
                        queue.poll();
                }
                time++;
                if (queue.isEmpty() && temp.size() == 0)
                    break;
                i++;
            }
            for (int l: temp)
                queue.add(l);
        }
        return time;
    }
}

*/